### ISSUE TITLE
Non-ResX resources get different resource IDs on Windows compared to C#

### ISSUE THREAD

**Issue by nosami** [Microsoft/visualfsharp#922]
Folder names aren't respected when generating resource ids passed to the `--resource` parameter of fsc.exe
- Create a new F# Console project
- Add a folder called "AFolder"
- Add a text file to AFolder called "AResourceFile.txt"
- Set the build action on that text file to embedded resource

``` fs
open System
open System.Reflection

[<EntryPoint>]
let main argv = 
    Assembly.GetExecutingAssembly().GetManifestResourceNames()
    |> Array.iter (Console.WriteLine) 
    0 // return an integer exit code
```

outputs 

```
AResourceFile.txt
```

Performing the same steps in a C# project results in 

```
csconsoleresource.AFolder.AResourceFile.txt
```

Note too that it isn't possible to override the resource id by using

``` xml
<LogicalName>My.Custom.Identifier</LogicalName>
```

in the msbuild file.

See https://bugzilla.xamarin.com/show_bug.cgi?id=37971 and https://github.com/fsharp/fsharp/issues/50 for further background.

**Comment by mexx** [comment]
FYI when using the `Resource` build action, folders are correctly managed, albeit all packaged into the global resource, so some extra work is necessary to access them. You can look at an example of my [GlobalResourceFileSystem](https://gist.github.com/mexx/aef547bdc4f07663fe1b) for OWIN.

**Comment by dsyme** [comment]
I seem to remember some subtle behaviour difference between the F# open edition and the Visual F# Tools here, due to some bug in Mono.  It's possible that bug has now been fixed

@nosami could you check if this problem exists in Visual F# Tools, or just Mono F#, or both? thanks

**Comment by nosami** [comment]
The two bugs are present in both. xbuild delegates the task of generating resource Ids to FSharp.Build.dll which appears to come from this repo. 

Note that Xamarin Studio recently started using code from MSBuild to build the code, so there is very little difference in the build process any more (if any). At some time in the past, it used it's own mechanism (which does actually do the correct thing here)

I created a repro solution [here](https://github.com/nosami/fsconsoleresource/tree/master/fsconsoleresource)

**Comment by dsyme** [comment]
@nosami There's a long analysis of this [here](https://github.com/fsharp/fsharp/pull/559#issuecomment-207575622), and you're right that the two bugs described above do exist in the Windows implementation.

I can't for the life of me work out where ``LogicalName`` is consumed even by C# MSBuild on Windows, and why it would not be being applied for F#.  If someone can point me in the right direction I'd be really grateful.  This stuff is just impenetrable.

BTW am I right in thinking that the command ``msbuild``  is now available in the Mono linux packages?

**Comment by dsyme** [comment]
@nosami I've clarified that for Windows this only applies to non-ResX resources.  For Linux, the problem also applies to ResX resources

Applying a fix for this on Windows would be a breaking change - do you see if there's a way to fix this without a breaking change?

**Comment by dsyme** [comment]
@nosami I've put the specific issue about LogicalName here: https://github.com/Microsoft/visualfsharp/issues/1050.  That's probably the easiest thing to fix.

**Comment by nosami** [comment]
Looking through the Xamarin code, it looks as if `LogicalName` is used by custom msbuild tasks rather than msbuild itself.

It will be a breaking change for Mono too, so not sure what the answer is here. It's quite a big issue at the moment for us, because iOS uses the same resource ID for different icon sizes and we have no workaround. I'd rather that you adopted the correct behaviour, but would understand if you can't. 

Being able to use `LogicalName` to override the generated name would at least give our end users a workaround. `LogicalName` is what we set when changing the resource ID in the IDE.

`msbuild` is currently in the bleeding edge mono builds so I'd expect it be adopted soon, probably the next public release. 

Thanks!

**Comment by dsyme** [comment]
> It will be a breaking change for Mono too

Is that for XBuild or Mono more generally?

**Comment by nosami** [comment]
It will only be a breaking change for us if a developer references the resource ID as a string in their code (and easy to fix).

**Comment by nosami** [comment]
> Is that for XBuild or Mono more generally?

XBuild or MSBuild delegates the job of generating resource IDs to the same msbuild task (that references this dll)

This issue affects F# only (xbuild or msbuild, Windows or Mono). C# is fine.

**Comment by dsyme** [comment]
> I'd rather that you adopted the correct behaviour, but would understand if you can't. 

I believe we'd be happy to fix this, but I looked around and really couldn't find where the LogicalName information was consumed.  It's likely an easy fix but I don't have any leads right now.  If you find something please send a link!

**Comment by nosami** [comment]
I've seen code, but it's in currently private Xamarin repos :)

All it does is use the name specified inside `<LogicalName>resourceID</LogicalName>` if it exists, otherwise uses generates the ID from the filename (including projectName and folder!)

**Comment by dsyme** [comment]
OK thanks.  Please send any snippets you have. However I couldn't determine where the C# open source MSBuild code for this resides - that's really what we should be following.

Where would the corresponding fix go for F#?  In Microsoft.FSharp.targets?

**Comment by nosami** [comment]
The mono code for this is [here](https://github.com/mono/mono/blob/ef407901f8fdd9ed8c377dbec8123b5afb932ebb/mcs/class/Microsoft.Build.Tasks/Microsoft.Build.Tasks/ManagedCompiler.cs#L93-L101) 

ItemSpec comes from [this function](https://github.com/mono/mono/blob/ef407901f8fdd9ed8c377dbec8123b5afb932ebb/mcs/class/Microsoft.Build.Tasks/Microsoft.Build.Tasks/CreateManifestResourceName.cs#L92)

Haven't seen any equivalent MS code, but this looks to generate the same ID as C# on MS .Net

**Comment by nosami** [comment]
I did find where this this fix should go once, but it's escaping me now - I should have linked to it when I created the issue!

**Comment by ForNeVeR** [comment]
@KevinRansom looks like this issue was indented to be closed by commit f6049d54d464874559765ced35276978656a5e52, right?

**Comment by KevinRansom** [comment]
@ForNeVeR yes it seems like it.
